libs/bench/include/nil/crypto3/bench/benchmark.hpp

```diff
@@ -0,0 +1,163 @@
+    const std::array<std::size_t, sizeof...(A)>& strides,
+    std::size_t i,
+    std::index_sequence<I...>
+) -> std::tuple<typename A::value_type&...>
+{
+    return std::tuple<typename A::value_type&...>(
+        std::get<I>(samples)[i * strides[I] % SAMPLES_COUNT]...
+    );
+}
+
+
+template<typename... A>
+auto make_slice(
+    std::tuple<std::vector<typename A::value_type>...>& samples,
+    const std::array<std::size_t, sizeof...(A)>& strides,
+    std::size_t i
+) -> std::tuple<typename A::value_type&...>
+{
+    return make_slice_impl<A...>(
+        samples, strides, i, std::index_sequence_for<A...>{}
+    );
+}
+
+
+template<typename...A, typename F >
+void run_benchmark(std::string const& name, F && func)
+{
+    using duration = std::chrono::duration<double, std::nano>;
+
+    auto samples = allocate_samples<A...>();
+    auto strides = calculate_strides<A...>();
+
+    auto run_batch = [&] (std::size_t batch_size) {
+        for(std::size_t i = 0; i < batch_size; ++i) {
+            auto args = make_slice<A...>(samples, strides, i);
+            /* volatile hints to compiler that it has important side effects
+             * and call should not be optimized out */
+            volatile auto r = std::apply(func, args);
+            (void) r;
+        }
+    };
+
+    auto run_at_least = [&] (duration const& dur) {
+        std::size_t WARMUP_BATCH_SIZE = 1000, total_runs = 0;
+        auto start = std::chrono::high_resolution_clock::now();
+        while (std::chrono::high_resolution_clock::now() - start < dur) {
+            run_batch(WARMUP_BATCH_SIZE);
+            total_runs += WARMUP_BATCH_SIZE;
+        }
+        return total_runs;
+    };
+
+    std::size_t MEASUREMENTS = 100;
+    duration WARMUP_DURATION = std::chrono::seconds(3);
+
+    std::size_t BATCH_SIZE = 1 + run_at_least(WARMUP_DURATION)/MEASUREMENTS/10;
+
+    std::vector<double> durations(MEASUREMENTS);
+    for(std::size_t m = 0; m < MEASUREMENTS; ++m) {
+        auto start = std::chrono::high_resolution_clock::now();
+        run_batch(BATCH_SIZE);
+        auto finish = std::chrono::high_resolution_clock::now();
+        durations[m] = (finish - start).count()*1.0 / BATCH_SIZE;
+    }
+
+    std::sort(durations.begin(), durations.end());
+
+    // discard top 20% outliers
+    durations.resize(MEASUREMENTS * 0.8);
+
+    double median = durations[durations.size()/2];
+    double mean = 0, stddiv = 0;
+
+    for(auto &dur : durations) {
+        mean += dur;
+        stddiv += dur*dur;
+    }
+
+    mean /= durations.size();
+    // stddiv^2 = E x^2 -  (E x)^2
+    stddiv = sqrt(stddiv / durations.size() - mean * mean);
+
+    // https://support.numxl.com/hc/en-us/articles/115001223503-MdAPE-Median-Absolute-Percentage-Error
+    for(auto &dur : durations) {
+        dur = (dur - median) / dur;
+        if ( dur < 0 ) {
+            dur = -dur;
+        }
+    }
+    std::sort(durations.begin(), durations.end());
```

Comment:
Probably you don't need to sort the durations again after normalization.

Reply:
no, we need. After APE is calculated, we need to get median. The first half is negative, the least values are in the middle and the last part is positive. and we get abs values. 
Consider this (sorted) array: 1 2 3 4 5. It has median = 3. When calculating MdAPE, we get this array:
`-2 -0.5 0 0.25 0.4`
We take absolute values and take median of that array. Yes, it needs to be sorted.